goaviatrix/version.go

```diff
@@ -56,58 +56,86 @@ func (av *AviatrixVersion) String(includeBuild bool) string {
 		return fmt.Sprintf("%d.%d.%d", av.Major, av.Minor, av.Build)
 	}
 	return fmt.Sprintf("%d.%d", av.Major, av.Minor)
-
 }
 
-func (c *Client) Upgrade(version *Version, upgradeGateways bool) error {
-	if version.Version == "" {
-		return errors.New("no target version is set")
-	}
-	Url, err := url.Parse(c.baseURL)
-	if err != nil {
-		return errors.New(("url Parsing failed for upgrade") + err.Error())
+// AsyncUpgrade will upgrade controller asynchronously
+func (c *Client) AsyncUpgrade(version *Version, upgradeGateways bool) error {
+	form := map[string]string{
+		"CID":    c.CID,
+		"caller": "ui", // indicates an async command
 	}
-	var action string
-	upgradeController := url.Values{}
-	upgradeController.Add("CID", c.CID)
 	if upgradeGateways {
-		action = "upgrade"
-		upgradeController.Add("action", action)
+		form["action"] = "upgrade"
 		if version.Version != "latest" {
-			upgradeController.Add("version", version.Version)
+			form["version"] = version.Version
 		}
 	} else {
-		action = "upgrade_platform"
-		upgradeController.Add("action", action)
-		upgradeController.Add("gateway_list", "")
-		upgradeController.Add("software_version", version.Version)
+		form["action"] = "upgrade_platform"
+		form["gateway_list"] = ""
+		form["software_version"] = version.Version
 	}
-
-	for i := 0; ; i++ {
-		Url.RawQuery = upgradeController.Encode()
-		resp, err := c.Get(Url.String(), nil)
+	resp, err := c.Post(c.baseURL, form)
+	if err != nil {
+		return fmt.Errorf("HTTP POST %s failed: %v", form["action"], err)
+	}
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(resp.Body)
+	requestID, err := strconv.Atoi(buf.String())
+	if err != nil {
+		// Could not decode as integer, so something went wrong
+		// try decoding as JSON to get an error message.
+		var data APIResp
+		err = json.Unmarshal(buf.Bytes(), &data)
 		if err != nil {
-			return fmt.Errorf("HTTP Get %s failed: %v", action, err)
+			// Could not decode as JSON either, something is very wrong.
+			return fmt.Errorf("Decode %s failed: %v\n Body: %s", form["action"], err, buf.String())
 		}
-		var data UpgradeResp
-		buf := new(bytes.Buffer)
+		return fmt.Errorf("rest API %s POST failed to initiate async action: %v", form["action"], data.Reason)
+	}
+	// Use the requestID to poll until upgrade is finished
+	form = map[string]string{
+		"action": "check_upgrade_status",
+		"CID":    c.CID,
+		"id":     strconv.Itoa(requestID),
+		"pos":    "0",
+	}
+	backendURL := fmt.Sprintf("https://%s/v1/backend1", c.ControllerIP)
+	const maxPoll = 60
+	sleepDuration := time.Second * 30
```

Comment:
Maybe sleep duration of 10 seconds and maxpoll of 180 would be slightly more responsive

Reply:
Good idea, updated.